### PR TITLE
fix(huffman): Handle edge cases and improve error handling

### PR DESCRIPTION
**Pull request by jesper-olsen**
## Description

This PR improves the robustness of the Huffman encoding implementation by properly handling edge cases that previously caused undefined behavior or panics.

### Summary of changes:
- Changed `HuffmanDictionary::new()` to return `Option<Self>` for safer API design
- Added proper handling for empty alphabets (returns `None` instead of panicking)
- Added special case handling for single-symbol alphabets, which previously failed during tree construction
- Improved error handling in `decode()` by replacing `unwrap()` calls with `?` operator
- Added `#[inline(always)]` optimization for the frequently-called `get_bit()` method
- Added comprehensive documentation with usage examples
- Added tests for edge cases (empty and single-symbol inputs)

The Huffman encoding algorithm is a well-known lossless data compression algorithm. More details: https://en.wikipedia.org/wiki/Huffman_coding

**Note**: During development, I observed intermittent heap corruption errors in the test suite that occur even without these changes, suggesting pre-existing issues in the codebase.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [ ] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [ ] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `CONTRIBUTING.md` and my code follows its guidelines.

Please make sure that if there is a test that takes too long to run ( > 300ms), you `#[ignore]` that or
try to optimize your code or make the test easier to run. We have this rule because we have hundreds of
tests to run; If each one of them took 300ms, we would have to wait for a long time.

### CODE REVIEW

File: src/machine_learning/loss_function/kl_divergence_loss.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ pub fn kld_loss(actual: &[f64], predicted: &[f64]) -> f64 {
     let loss: f64 = actual
         .iter()
         .zip(predicted.iter())
-        .map(|(&a, &p)| ((a + eps) * ((a + eps) / (p + eps)).ln()))
+        .map(|(&a, &p)| (a + eps) * ((a + eps) / (p + eps)).ln())
```
Comment (vil02) [inline src/machine_learning/loss_function/kl_divergence_loss.rs:19]: This seems not relevant. Please revert it here and create another tiny PR - we will merge it.

**Review by vil02 — CHANGES_REQUESTED** [review]
The change looks good. One minor thing below.
